### PR TITLE
Structured docs

### LINKED ISSUE #658: Improve command and bundle help documentation.
# Problem

Currently, the amount of documentation related metadata that we store for commands and bundles is limited to an arbitrary text string for each command that is expected to contain the documentation for using the command. This limits the amount of help that we can provide to users and our ability to format that help information in a meaningful way.

Let's remodel of the documentation related parts of bundle configuration and improve the `operable:help` command to provide more useful and consistently structured results

# Proposed Config Changes

All keys are optional unless specifically called out as required.

## Bundles

Add the following keys to the top-level of the bundle configuration:

- **`description`** - Short description of the bundle that can be used by `operable:help` to describe the functions that the bundle provides. (required)
- `long_description` - Longer description that is used when requesting details about the bundle. 
- `author` - String to be included in the AUTHOR section of the bundle help.
- `homepage` - String to be included in the HOMEPAGE section of the bundle help.

## Commands

Add the following optional keys to the configuration of each command:

- `description` - Short description of the command, used when listing the commands in a bundle and in the detailed man page for the command.
- `long_description` - Longer description that is used at the top of the detailed help page for the command.
- `examples` - Examples of use of the command with various options.
- `notes` - Free form NOTES section that will appear at the end of the help page, just above AUTHOR and HOMEPAGE information.
- `arguments` - String to be added to the end of the generated synopsis to represent positional arguments. This is kind of a hack to avoid having to make positional arguments as rigid as Options, so that users can represent things like "instance1 .. instanceN" in their documentation. **I'm not sure if this is the right call for arguments or not. Feedback and ideas are extra requested here.**

# Proposed Help Command Behavior

- `operable:help` - List installed bundles with a short description of each.
- `operable:help <bundle>` - Display long description of a specific bundle and a list of the commands that are included.
- `operable:help <bundle>:<command>` - Display a man page for an individual command.

#### `operable:help`:
```
Available bundles:
  
   * operable - Cog administration and support utilities.
   * mist - Amazon Web Services management tools.

Disabled bundles:

   * format - Support commands for formatting pipeline output.
   * circle - CircleCI integration

To learn more about a specific bundle and the commands available within
it, you can use `operable:help <bundle name>`.
```

#### `operable:help <bundle>`:
```
NAME:
  mist - Amazon Web Services management tools.

DESCRIPTION:
  mist is  collection of commands for managing resources in Amazon
  Web Services. Currently a subset of the EC2, S3, and VPC services
  are included.

  These commands all require a --region option to be included to identify
  the AWS region to operate against. You can select a default region 
  by defining a MIST_DEFAULT_REGION variable in your relay's dynamic
  command configuration.

COMMANDS: 
  * mist:ec2-create - Create an EC2 instance
  * mist:ec2-destroy - Destroy an EC2 instance
  * mist:ec2-find - Find EC2 instances
  * mist:ec2-state - ...
  * mist:ec2-tags - ...
  * mist:keypairs-list - ...
  * mist:s3-bucket-acl - ...
  * mist:s3-bucket-files - ...
  * mist:s3-buckets - ...
  * mist:vpc-list - ...

  For details about a specific command, you can use the  `operable:help <command>`, 
  for example `operable:help mist:ec2-find`.

AUTHOR:
  Kevin Smith <kevin@example.com>

HOMEPAGE: 
  https://github.com/cog-bundles/mist/
```

#### `operable:help <bundle>:<command>`:

Note: The `mist:ec2-find` command doesn't actually have any positional
arguments but I include an `instance_id` argument in the synopsis just
to illustrate how named positional arguments may be handled when 
generating the synopsis.

```
NAME
  mist:ec2-find - Find EC2 instances

DESCRIPTION:
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam molestie augue at diam egestas, 
  ac fringilla orci pulvinar. Pellentesque habitant morbi tristique senectus et netus et malesuada
  fames ac turpis egestas. Nulla id nibh egestas, fermentum magna eu, hendrerit risus. 

SYNOPSIS:
  mist:ec2-find --region=<region> [--state <state>] [--tags <tags>]
                [--ami <ami>] [--return <field_list>] instance_id

REQUIRED OPTIONS:
  --region
      The Amazon Web Services region to find instances in.

OPTIONS:
  --state <state>
      Instance state, e.g. "running", "stopped", etc.
      
  --tags <tag>
      List of tags to find. 

      Tags consist of Key=Value pairs. Multiple tags should be separated
      by commas. 
  --ami <ami>
      EC2 AMI ID to find.

  --return <return>
      A comma separated list of fields to return in the results.

      Available fields: 
        id, pubdns, privdns, state, keyname, ami, kernel, vpc, 
        pubip, privip, az, tags

EXAMPLES:
  Return all instances in the us-east-1 region:
  
    mist:ec2-find --region us-east-1 

  Return instances in the us-east-1 region with the Environment=staging
  tag set:

    mist:ec2-find --region us-east-1 --tags Environment=staging

  Return the instance ID and public IP address for running instances 
  in the us-east-1 region:

    mist:ec2-find --region us-east-1 --state running --return id,pubip

NOTES:
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam molestie augue at diam egestas, 
  ac fringilla orci pulvinar. Pellentesque habitant morbi tristique senectus et netus et malesuada
  fames ac turpis egestas. Nulla id nibh egestas, fermentum magna eu, hendrerit risus.
 
AUTHOR:
  Kevin Smith <kevin@example.com>

HOMEPAGE: 
  https://github.com/cog-bundles/mist/
```

### PR DESCRIPTION
**Pull request by vanstee**
⚠️ Waiting on new templates to potentially use those instead of returning raw text.

Closes #658

### CODE REVIEW

File: lib/cog/bundle/embedded.ex [hunk 1]
```diff
@@ -63,6 +63,8 @@ defmodule Cog.Bundle.Embedded do
     Config.gen_config(Cog.Util.Misc.embedded_bundle,
                       "Core chat commands for Cog",
                       version,
+                      "Operable <support@operable.io>",
```
Comment (christophermaier) [inline lib/cog/bundle/embedded.ex]: This is cool, but we should probably simplify our config generation code soon; this is literally the only bundle that uses it, so we don't need to take so many pains to keep it general.

File: lib/cog/command/gen_command/base.ex [hunk 2]
```diff
@@ -173,8 +178,10 @@ defmodule Cog.Command.GenCommand.Base do
     attr_value(module, :command_name)
   end
 
-  def description(module),
-    do: attr_value(module, :description)
+  for attr <- [:description, :long_description, :examples, :notes, :arguments, :subcommands] do
```
Comment (christophermaier) [inline lib/cog/command/gen_command/base.ex]: Guess you might as well pull `:command_name`, `:bundle_name`, `:rules`, and `:permissions` into this as well.

File: test/controllers/v1/bundles_controller_test.exs [hunk 3]
```diff
@@ -64,24 +64,33 @@ defmodule Cog.V1.BundlesControllerTest do
...
+    %{"errors" => errors} = Poison.decode!(conn.resp_body)
+
+    assert errors == ["Invalid config file.",
+                      "Error near #/cog_bundle_version: cog_bundle_version 2 is not supported. Please update your bundle config to version 4."]
+  end
+
   test "accepts an upgradable config", %{authed: requestor} do
-    config_path = make_config_file("old_config.yaml", contents: old_config)
-    upload = %Plug.Upload{path: config_path, filename: "old_config.yaml"}
+    config_path = make_config_file("upgradable_config.yaml", contents: upgradable_config)
```
Comment (christophermaier) [inline test/controllers/v1/bundles_controller_test.exs]: This whole test might just go away instead... the idea of "upgradable configs" was from when we'd make automatic changes to go from v2 to v3. We don't actually have that code anymore.

File: lib/cog/command/gen_command/base.ex (hunk 2 above)
Comment (vanstee) [inline lib/cog/command/gen_command/base.ex]: Tried to make this change, but there's some weird macro magic getting in the way. Couldn't figure out exactly what was causing it so I'm just going to leave this as is for now.

**Pull request merged**